Updates to Research Activity time calculations. (Plus minor fixes)

beelitzs opened the pull request:
This update adds the level modifier system rather than each run of a ME or PE job always taking the same amount of time regardless of level.

- For now I've left a static divisor using the same value as the original code to keep the values realistic. The discrepancy is caused by the difference in the base value between the SDE used by Crucible compared to the current live SDE.
- Formula used pulled from https://eve-industry.org/export/IndustryFormulas.pdf which is whats currently used on live however I believe the same system was used back in Crucible

- Additionally added a minor bug fix for #182 where floats >0 but <1 were creating stacks of 0 ore in cargo.

beelitzs commented:
@jdhirst This PR is still WIP however I wanted to created it early to hopefully generate some feedback as this is my first experience with C++.

jdhirst commented:
!oktotest

Almamu commented:
Don't mind me, just ensuring our jenkins instance properly builds PRs... !oktotest

Almamu commented:
!oktotest

Almamu commented:
!oktotest again -.-'

jdhirst commented:
Looks good to me, I think these rate changes make sense; @Almamu what do you think?

Almamu commented on src/eve-common/utils/EvEMath.cpp:
```diff
@@ -65,14 +65,18 @@ int32 EvEMath::RAM::ProductionTime(uint32 BaseTime, float bpProductivityModifier
 }
 
-int32 EvEMath::RAM::ME_ResearchTime(uint32 BaseTime, uint8 MetallurgyLevel, float SlotModifier/*1*/, float ImplantModifier/*1*/ )
+const int LEVELMODIFIERS [11] = { 0, 105, 250, 595, 1414, 3360, 8000, 19000, 45255, 107700, 256000 }; // This probably needs to go to a more appropriate location
```
This should go into it's own function with a bounds check to ensure we do not read out of the array as the "Runs" parameter comes directly from the client's call, which is easily forged.

beelitzs replied:
Yes that right, in its current state it works well up to level 10 and then for 11 and above it returns zero which the client interprets as an instantaneous job.
My intention to fix this is to add a check earlier on that feeds back to the client (I'm still working out how to do this) when the current level + runs > 10 which will fix the current issue of being able to enter too many runs for a research job and by extension solve the boundary issue at the same time.

Another thing I will also need to make allowances for is that wile ME goes up by 1 to a maximum of 10, PE goes up by 2 up to a maximum of 20 so I'll need to ensure the code can handle that.

Thoughts?

Almamu replied:
It's usually a good idea to have a bounds check even if you validate the client's data, this way you ensure that if someone forgets to do it that your app doesn't crash or read garbage data, which might or might not be what you expect. With the changes it looks good to me now. Nice work ;)

Almamu commented:
I've added a comment on a small improvement that could be done, looks fine to me after that's fixed.

beelitzs commented:
- Added a client error for when the user attempts to create a research job that would result in the BP leveling beyond 10.
- I was originally using the wrong method to get the current ME/PE value causing the values to be incorrect, fixed.
- Moved the LevelModifier calc into its own method and added a boundary check as per @Almamu 's comment.

I'm much more happy with it now with the minor exception of the "RamActivityInvalid" I'm currently throwing (there should be a more appropriate one to use in the client however I'm unable to find an example in RamMethods.h)
The PE issue I was worried about doesn't seem to be an issue at all and in fact it may have even been introduced post Crucible so will leave as is since its functioning fine.

Almamu commented:
!oktotest

Almamu commented:
> I'm much more happy with it now with the minor exception of the "RamActivityInvalid" I'm currently throwing (there should be a more appropriate one to use in the client however I'm unable to find an example in RamMethods.h)

There's situations where the client is vague on purpose, like for example in the case of you being able to modify the parameters the server won't say "hey, your parameters are wrong". To me the PR looks good, It's checking the build process now but unless @jdhirst has anything to say, I think it can be merged.

jdhirst commented:
Yeah I think we are good to go for this, great work!

Almamu commented:
Can one of the admins verify this patch?
